### PR TITLE
retry poll failures with exponential backoff (see #122)

### PR DESCRIPTION
**Pull request by breunigs**
As per https://github.com/zendesk/racecar/issues/122#issuecomment-530777250 , let's keep to the existing promises. 

It doesn't fix the specific issue mentioned in #122, though.

### CODE REVIEW

**Comment by dasch** [comment]
Awesome!

I do think, however, that we need to respect `timeout_ms`. How about stopping retries once that time interval has passed? Otherwise we could be looking at very long latencies for the call.

File: lib/racecar/consumer_set.rb [hunk 1]
```diff
@@ -12,21 +14,21 @@ def initialize(config, logger)
 
     def poll(timeout_ms)
       maybe_select_next_consumer
```
Comment (dasch) [inline lib/racecar/consumer_set.rb]: I’m thinking we set a deadline here and check it before retrying.

**Comment by breunigs** [comment]
You bring up a valid point, but I'm wondering about decent default values. If the racecar runner is used, then the value here is `config.max_wait_time` which is:
```
    desc "How long to allow the Kafka brokers to wait before returning messages"
    float :max_wait_time, default: 1 # seconds
```
This value is also used for batch polling, so the 1 second seems a decent trade-off between ability to batch and lag. To avoid exceeding the limit too much, I'll amend the code to pass the remaining time budget instead of the full budget for every poll. 

With the current settings there'll be at most 1 retry, which seems undesirable. Adjusting the calculation to e.g.
```ruby
sleep (2**(try-1))/10
```
will give sleeps of 100ms, 200ms, 400ms, 800ms, …, 51.2s. Assuming a failing poll doesn't take too long, there'd be at least 4 tries, which seems reasonable. Well, 5, if we decide the additional retry is worth more than exceeding the time limit. I'd err on early failure here, though.

**Comment by dasch** [comment]
I think that's a great approach :+1:

**Comment by breunigs** [comment]
Note that the code right now has a race condition:
1. has enough remaining time to attempt retry
2. something in `rescue` block is slow (e.g. logger being throttled)
3. next try re-evaluates remaining time and simply returns `nil` 

I'll see if I can find a decent fix

**Comment by dasch** [comment]
@breunigs are you happy with the current version? Ready to merge?

**Comment by breunigs** [comment]
Yes, I was about to write a "ready for review" comment to clean up any misunderstandings, because I consider this good to go now.

**Comment by dasch** [comment]
Thanks!

**Pull request merged**